10-JavaScript/js-ecosystem.md

```diff
@@ -0,0 +1,153 @@
+
+  webpack is the "main" tool in most modern JS build systems (there are, as always, alternatives). While it bills itself as a "module bundler", its role is actually closer to "build system manager/orchestrator". Most other build tools are added into the build system by including them into your webpack configuration files.
+
+  webpack is also responsible for locating all the source files (referred to as "modules") in your project, sending those files to the appropriate tools for processing, and packaging all of the processed modules into a "bundle" of one or more files fit for deployment to end users.
+
+  Given its role as the manager of the whole build system, webpack itself is actually a very generic tool that can do just about anything depending on its exact configuration and installed plugins. In the latest version of webpack as of this writing, version 4, there is a strong emphasis on allowing webpack to be used without any configuration by ensuring that the default options for everything are usable.
+
+* [Babel](https://babeljs.io/) - JavaScript compiler
+
+  Babel's job is to take JavaScript code written in a [newer version of JavaScript](https://babeljs.io/learn-es2015/) (known as ES6 or ES2015) and convert it to equivalent code in an older version of JavaScript (generally ES5) which has more universal support in web browsers.
+
+  Babel is often used as the first stage by webpack when it processes your JS code files, followed possibly by minification and other transformations.
+
+* [webpack-dev-server](https://github.com/webpack/webpack-dev-server)
+
+  When we develop our client-side JS applications it's necessary to actually access the application through our browser locally, so that we can see the behavior of our code. One way to accomplish this might be to directly open the HTML file in our browser, as we did for our jQuery projects and when learning HTML and CSS.
+
+  However, this approach has some downsides. Specifically, the browser is loading a _file_ URL which looks like `file:///Users/username/some/directory/to/a/file.html`, which is treated differently from an _http(s)_ URL. For large or potentially complex projects it's important to mimic the actual end-user experience as closely as possible so we really want the browser to access our site through HTTP just as it would any other website on the internet.
+
+  In order to accomplish this we need to run a web server which will actually provide our HTML, CSS, JavaScript, and other files from a given URL (generally `localhost:3000` or similar). That is the purpose of the webpack-dev-server tool, specifically built to integrate with a webpack-based build system.
+
+  In very complex applications the developers might modify this "dev server" to have custom behavior akin to an API, but for our purposes we just need to run it with the default setup. In this case the dev server can most closely be compared to running `rails server` for our Rails projects.
+
+## Create React App
```

Comment:
Starting here, this section is for the following learning goal:
- Explore a complex build system setup by "ejecting" the build configuration from a `create-react-app` project

How would you feel if before this header, we had the following top-level header above it:
`# Explore A Complex Build System Setup through Create React App`

?

That might help structure a student's expectation for what to get out of this section